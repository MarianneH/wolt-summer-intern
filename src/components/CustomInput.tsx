import styled from "styled-components";
import { RiPinDistanceLine } from "react-icons/ri";
import { AiOutlineEuroCircle } from "react-icons/ai";
import { AiOutlineShoppingCart } from "react-icons/ai";
import { TbCalendarTime } from "react-icons/tb";
import { useState } from "react";

const StyledInput = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 1rem;
  position: relative;
  width: 100%;
  font-size: 1rem;
  & input {
    width: 100%;
    box-sizing: border-box;
    background: rgb(255, 255, 255);
    color: rgb(32, 33, 37);
    padding: 0.75rem 0.875rem;
    padding-top: 1.25rem;
    border: 2px solid rgba(32, 33, 37, 0.12);
    border-radius: 0.5rem;
    padding-inline-start: 3.125rem;
    line-height: 1.4;
    font-size: 1rem;
    transition: border 0.2s linear 0s;
    &:focus {
      + label {
        transform: translateY(-0.6875rem) scale(0.75);
      }
    }
    &:hover {
      border-color: rgb(0, 157, 224);
      box-shadow: rgb(0, 157, 224) 0px 0px 0px 0.0625rem;
      transition: border 0.2s linear 0s;
    }
  }
  & input[type="datetime-local"]::-webkit-calendar-picker-indicator {
    background: transparent;
    bottom: 0;
    color: transparent;
    cursor: pointer;
    height: auto;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    width: auto;
  }
  & label {
    line-height: 1.5rem;
    color: rgba(32, 33, 37, 0.64);
    transform-origin: left center;
    position: absolute;
    left: 3.25rem;
    transition: transform 0.15s ease-in-out 0s;
    pointer-events: none;
    &.active {
      transform: translateY(-0.6875rem) scale(0.75);
    }
  }
  & div {
    pointer-events: none;
    cursor: text;
    width: 1.5rem;
    height: 1.5rem;
    position: absolute;
    display: flex;
    align-items: center;
    justify-content: center;
    left: 1rem;
    right: auto;
    color: rgba(32, 33, 37, 0.64);
    & svg {
      width: 2rem;
      height: 2rem;
    }
  }
`;

interface CustomInputProps {
  version: "distance" | "amount" | "time" | "cartValue";
  inputRef: React.RefObject<HTMLInputElement>;
}

export const CustomInput: React.FC<CustomInputProps> = ({
  version,
  inputRef,
}) => {
  const [hasValue, setHasValue] = useState<boolean>(false);
  function handleInputChange(e: any) {
    if (e.target.value) {
      setHasValue(true);
    } else {
      setHasValue(false);
    }
  }

  return (
    <StyledInput>
      <input
        onChange={handleInputChange}
        name={version}
        type={version === "time" ? "datetime-local" : "number"}
        min="0"
        step={version === "cartValue" ? "0.01" : ""}
        required
        ref={inputRef}
      ></input>
      <label className={hasValue || version === "time" ? "active" : ""}>
        {version === "distance" && "Delivery Distance in m"}
        {version === "amount" && "Number of items"}
        {version === "time" && "Order time"}
        {version === "cartValue" && "Cart Value in â‚¬"}
      </label>
      <div>
        {version === "distance" && <RiPinDistanceLine />}
        {version === "amount" && <AiOutlineShoppingCart />}
        {version === "time" && <TbCalendarTime />}
        {version === "cartValue" && <AiOutlineEuroCircle />}
      </div>
    </StyledInput>
  );
};

export default CustomInput;
