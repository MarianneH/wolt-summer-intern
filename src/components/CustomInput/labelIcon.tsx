import { AiOutlineEuroCircle, AiOutlineShoppingCart } from "react-icons/ai";
import { RiPinDistanceLine } from "react-icons/ri";
import { TbCalendarTime } from "react-icons/tb";

export const labelIcon = {
  distance: {
    label: "Delivery Distance in m",
    icon: <RiPinDistanceLine />,
  },
  amount: {
    label: "Number of Items",
    icon: <AiOutlineShoppingCart />,
  },
  time: {
    label: "Order Time",
    icon: <TbCalendarTime />,
  },
  cartValue: {
    label: "Cart Value in â‚¬",
    icon: <AiOutlineEuroCircle />,
  },
};
