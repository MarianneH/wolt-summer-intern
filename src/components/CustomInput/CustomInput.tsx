import React, { useState } from "react";
import { StyledInput } from "./CustomInput.styled";
import { labelIcon } from "./labelIcon";

interface CustomInputProps {
  version: "distance" | "amount" | "time" | "cartValue";
  inputRef: React.RefObject<HTMLInputElement>;
}

export const CustomInput: React.FC<CustomInputProps> = ({
  version,
  inputRef,
}) => {
  const [hasValue, setHasValue] = useState<Boolean>(false);

  // check if input has value to handle label position
  function handleInputChange(e: React.ChangeEvent<HTMLInputElement>) {
    setHasValue(Boolean(e.target.value));
  }

  return (
    <StyledInput>
      <input
        onChange={handleInputChange}
        name={version}
        id={version}
        type={version === "time" ? "datetime-local" : "number"}
        min={version === "cartValue" ? "0.01" : "1"}
        step={version === "cartValue" ? "0.01" : "1"}
        required
        ref={inputRef}
        data-testid={version}
      />
      <label
        htmlFor={version}
        className={hasValue || version === "time" ? "active" : ""}
      >
        {labelIcon[version].label}
      </label>
      <div>{labelIcon[version].icon}</div>
    </StyledInput>
  );
};

export default CustomInput;
