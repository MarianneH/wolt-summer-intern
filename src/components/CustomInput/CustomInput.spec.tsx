import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
// import userEvent from "@testing-library/user-event";
import React from "react";
import CustomInput from "./CustomInput";

describe("CustomInput", () => {
  it("should render all properties correctly for version distance", () => {
    const { container } = render(
      <CustomInput
        version="distance"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should render input with type number if version is not time", () => {
    render(
      <CustomInput
        version="amount"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );

    expect(screen.queryByTestId("amount")?.getAttribute("type")).toEqual(
      "number"
    );
  });

  it("should render input with type datetime-local if version is time", () => {
    render(
      <CustomInput
        version="time"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );

    expect(screen.queryByTestId("time")?.getAttribute("type")).toEqual(
      "datetime-local"
    );
  });

  it("should render input with step 0.01 if version is cartValue", () => {
    render(
      <CustomInput
        version="cartValue"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );

    expect(screen.queryByTestId("cartValue")?.getAttribute("step")).toEqual(
      "0.01"
    );
  });

  it("should render input with step 1 if version is NOT cartValue", () => {
    render(
      <CustomInput
        version="amount"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );

    expect(screen.queryByTestId("amount")?.getAttribute("step")).toEqual("1");
  });

  it("should render input label with className active if value is not empty and version is not time", async () => {
    const user = userEvent.setup();

    render(
      <CustomInput
        version="amount"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );
    const label = screen.getByText("Number of Items");
    await user.click(screen.getByTestId("amount"));
    await user.keyboard("9");

    expect(label.classList.contains("active")).toBe(true);
  });

  it("should render input label with className active if version is time", async () => {
    render(
      <CustomInput
        version="time"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );
    const label = screen.getByText("Order Time");
    expect(label.classList.contains("active")).toBe(true);
  });

  it("should render input label without className active if version is distance and without input", async () => {
    render(
      <CustomInput
        version="distance"
        inputRef={{} as React.RefObject<HTMLInputElement>}
      />
    );
    const label = screen.getByText("Delivery Distance in m");
    expect(label.classList.contains("active")).not.toBe(true);
  });
});
