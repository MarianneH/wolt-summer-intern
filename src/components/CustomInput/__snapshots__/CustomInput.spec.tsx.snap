// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInput should render all properties correctly for version distance 1`] = `
<div>
  <div
    class="sc-bcXHqe fyNRQf"
  >
    <input
      data-testid="distance"
      id="distance"
      min="1"
      name="distance"
      required=""
      step="1"
      type="number"
    />
    <label
      class=""
      for="distance"
    >
      Delivery Distance in m
    </label>
    <div>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9.975 8.975a3.5 3.5 0 1 0-4.95 0L7.5 11.45l2.475-2.475zM7.5 14.278L3.61 10.39a5.5 5.5 0 1 1 7.78 0L7.5 14.28zM7.5 8a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm9 12.45l2.475-2.475a3.5 3.5 0 1 0-4.95 0L16.5 20.45zm3.89-1.06l-3.89 3.888-3.89-3.889a5.5 5.5 0 1 1 7.78 0zM16.5 17a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
